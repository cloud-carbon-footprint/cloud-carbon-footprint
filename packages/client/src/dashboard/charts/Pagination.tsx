/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React, {
  useState,
  useEffect,
  PropsWithChildren,
  ReactElement,
} from 'react'
import {
  ChevronLeft,
  ChevronRight,
  FirstPage,
  LastPage,
} from '@material-ui/icons'
import { IconButton, makeStyles } from '@material-ui/core'

interface UsePaginateData<T> {
  paginatedData: T[][]
  totalPages: number
}
export const usePaginateData: <T>(
  data: T[],
  pageSize: number,
) => UsePaginateData<T> = (data, pageSize) => {
  const paginatedData = []
  const newEntries = [...data]
  while (newEntries.length > 0) {
    const paginatedSubData = newEntries.splice(0, pageSize)
    paginatedData.push(paginatedSubData)
  }
  return {
    paginatedData,
    totalPages: paginatedData.length,
  }
}

const useStyles = makeStyles(() => {
  return {
    paginationContainer: {
      display: 'flex',
      width: '100%',
      justifyContent: 'flex-end',
      alignItems: 'center',
    },
  }
})

interface PaginateData<T> {
  data: T[]
  pageSize: number
}

export interface Page<T> {
  data: T[]
  page: number
}

export interface PaginationProps<T> extends PaginateData<T> {
  handlePage: (page: Page<T>) => void
}

export const Pagination: <T>(
  props: PropsWithChildren<PaginationProps<T>>,
) => ReactElement = ({ data, pageSize, handlePage }) => {
  const { paginationContainer } = useStyles()
  const [page, setPage] = useState(0)
  const { paginatedData, totalPages } = usePaginateData<typeof data[0]>(
    data,
    pageSize,
  )
  const visibleRows = `${page * pageSize + 1} - ${
    page * pageSize + paginatedData[page]?.length
  }`

  useEffect(() => {
    handlePage({ data: paginatedData[0] || [], page: 0 })
    setPage(0)
  }, [JSON.stringify(data)])

  const onPageChange = (newPage: number) => {
    setPage(newPage)
    handlePage({ data: paginatedData[newPage], page: newPage })
  }

  const lastPage = paginatedData.length - 1
  return data.length === 0 ? (
    <div aria-label="no-pagination-data" />
  ) : (
    <div className={paginationContainer}>
      <span style={{ color: '#ababab', fontWeight: 700, marginRight: '8px' }}>
        {visibleRows} of {data.length}
      </span>
      <IconButton
        color="primary"
        aria-label="first page"
        component="button"
        disabled={page === 0}
        onClick={() => onPageChange(0)}
      >
        <FirstPage />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="chevron-left"
        component="button"
        disabled={page === 0}
        onClick={() => onPageChange(page - 1)}
      >
        <ChevronLeft />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="chevron-right"
        component="button"
        disabled={page === totalPages - 1}
        onClick={() => onPageChange(page + 1)}
      >
        <ChevronRight />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="last page"
        component="button"
        disabled={page === lastPage}
        onClick={() => onPageChange(lastPage)}
      >
        <LastPage />
      </IconButton>
    </div>
  )
}
