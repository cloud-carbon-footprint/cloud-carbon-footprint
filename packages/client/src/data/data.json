[
    {
        "serviceEstimates": [
            {
                "co2e": 2,
                "cost": 0,
                "serviceName": "ebs",
                "timestamp": "2020-08-28T00: 00: 00.000Z",
                "wattHours": 0.5108360883835144
            },
            {
                "co2e": 1,
                "cost": 0,
                "serviceName": "s3",
                "timestamp": "2020-08-28T00: 00: 00.000Z",
                "wattHours": 0.00026440008612119945
            },
            {
                "co2e": 0,
                "cost": 2.330506638787581,
                "serviceName": "ec2",
                "timestamp": "2020-08-28T00: 00: 00.000Z",
                "wattHours": 50.5206029585591
            }
        ],
        "timestamp": "2020-08-28T00: 00: 00.000Z"
    },
    {
        "serviceEstimates": [
            {
                "co2e": 4,
                "cost": 0,
                "serviceName": "ebs",
                "timestamp": "2020-07-28T00: 00: 00.000Z",
                "wattHours": 0.23268313620749437
            },
            {
                "co2e": 4,
                "cost": 0,
                "serviceName": "s3",
                "timestamp": "2020-07-28T00: 00: 00.000Z",
                "wattHours": 0.0001680875709602132
            },
            {
                "co2e": 4,
                "cost": 1.9776640256171394,
                "serviceName": "ec2",
                "timestamp": "2020-07-28T00: 00: 00.000Z",
                "wattHours": 50.10627201163577
            }
        ],
        "timestamp": "2020-07-28T00: 00: 00.000Z"
    }
]