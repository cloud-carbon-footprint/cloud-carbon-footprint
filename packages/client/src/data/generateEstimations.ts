/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import moment from 'moment'
import { EstimationResult, serviceEstimate } from '../models/types'

const getRandomInt = (max: number): number => {
  return Math.floor(Math.random() * Math.floor(max))
}

const getRandomNumberInRange = (minValue: number, maxValue: number): number => {
  return Math.max(Math.random() * maxValue, minValue + Math.random())
}

const fakeServiceMap: () => { [key: string]: serviceEstimate } = () => {
  return {
    ebs: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'ebs',
      kilowattHours: Math.random(),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      usesAverageCPUConstant: false,
    },
    s3: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 's3',
      kilowattHours: Math.random() / 1000,
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-2',
      usesAverageCPUConstant: false,
    },
    ec2: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'ec2',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-west-1',
      usesAverageCPUConstant: false,
    },
    rds: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'rds',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-west-2',
      usesAverageCPUConstant: false,
    },
    lambda: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'lambda',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      usesAverageCPUConstant: false,
    },
    elasticache: {
      cloudProvider: 'AWS',
      accountName: `gcp account ${getRandomInt(5)}`,
      serviceName: 'elasticache',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      usesAverageCPUConstant: false,
    },
    computeEngine: {
      cloudProvider: 'GCP',
      accountName: `gcp account ${getRandomInt(5)}`,
      serviceName: 'computeEngine',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east1',
      usesAverageCPUConstant: false,
    },
    virtualMachines: {
      cloudProvider: 'AZURE',
      accountName: `azure account ${getRandomInt(5)}`,
      serviceName: 'virtualMachines',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'UK South',
      usesAverageCPUConstant: false,
    },
  }
}

const generateEstimations = (
  today: moment.Moment | Date,
  monthsBack: number,
  servicesToTest: string[] = [
    'ebs',
    's3',
    'ec2',
    'rds',
    'lambda',
    'elasticache',
    'computeEngine',
    'virtualMachines',
  ],
): EstimationResult[] => {
  const todayAsMoment: moment.Moment = moment(today)

  const estimationResults: EstimationResult[] = []

  for (let i = 0; i <= monthsBack; i++) {
    const timestamp: Date = todayAsMoment
      .clone()
      .utc()
      .subtract(i, 'M')
      .hours(0)
      .minutes(0)
      .seconds(0)
      .millisecond(0)
      .toDate()
    const estimationsForMonth: EstimationResult = {
      timestamp,
      serviceEstimates: servicesToTest.map((serviceToTest: string) => {
        return fakeServiceMap()[serviceToTest]
      }),
    }

    estimationResults.push(estimationsForMonth)
  }

  return estimationResults
}

export default generateEstimations
