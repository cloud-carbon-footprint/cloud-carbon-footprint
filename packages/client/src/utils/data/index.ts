/*
 * Â© 2021 Thoughtworks, Inc.
 */

export { fakeEmissionFactors, generateEstimations } from './generateEstimations'
export {
  mockData,
  mockDataWithUnknownsAWS,
  mockDataWithUnknownsGCP,
  mockDataWithHigherPrecision,
  mockDataWithSmallNumbers,
  mockDataWithLargeNumbers,
  mockRecommendationData,
  mockRecommendationDataWithUnknowns,
  mockEmissionsAndRecommendations,
  mockEmissionsAndRecommendationsWithUnknowns,
} from './mockData'
