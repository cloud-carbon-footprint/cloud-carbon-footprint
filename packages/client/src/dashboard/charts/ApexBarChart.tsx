/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent, useState, Fragment } from 'react'
import { useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { sumCO2ByServiceOrRegion } from '../transformData'
import { ApexChartProps } from './common/ChartTypes'
import { EmissionsRatios, chartBarCustomColors } from '../../models/types'
import { Page, Pagination } from './Pagination'
import ChartLegend from './ChartLegend'
import NoDataPage from '../NoDataPage'
import useRemoteEmissionService from '../client/EmissionFactorServiceHook'

const mapToRange = (
  value: number,
  in_min: number,
  in_max: number,
  out_min: number,
  out_max: number,
) => {
  return ((value - in_min) * (out_max - out_min)) / (in_max - in_min) + out_min
}

export interface Entry {
  x: string
  y: number
}

export const ApexBarChart: FunctionComponent<ApexChartProps> = ({
  data,
  dataType,
}) => {
  const [pageData, setPageData] = useState<Page<Entry>>({ data: [], page: 0 })
  const theme = useTheme()

  const {
    data: emissionsData,
    loading: _emissionsLoading,
  } = useRemoteEmissionService()

  const mainTheme = theme.palette.primary.main
  const darkTheme = theme.palette.primary.dark

  let customBarColors = [mainTheme]
  if (dataType === 'region' && !!emissionsData.length) {
    customBarColors = createCustomBarColors(pageData, emissionsData, mainTheme)
  }

  const barChartData = sumCO2ByServiceOrRegion(data, dataType)

  const dataEntries: { x: string; y: number }[] = Object.entries(barChartData)
    .filter((item) => item[1] > 0)
    .map((item) => ({
      x: item[0],
      y: item[1],
    }))
    .sort((higherC02, lowerCO2) => lowerCO2.y - higherC02.y)

  const smallestCO2E = dataEntries?.[dataEntries?.length - 1]?.y
  const largestCO2E = dataEntries?.[0]?.y
  const totalCO2EByDataType = dataEntries.reduce((acc, currentValue) => {
    return acc + currentValue.y
  }, 0)

  const pageSize = 10
  const minThreshold = 1
  const maxThreshold = 100
  const mappedDataEntries: Entry[] = dataEntries.map((entry) => {
    const yEntry = mapToRange(
      entry.y,
      smallestCO2E,
      largestCO2E,
      minThreshold,
      maxThreshold,
    )
    return { x: entry.x, y: isNaN(yEntry) ? maxThreshold : yEntry }
  })

  const toolbarOffset = {
    x: -120,
    y: dataType === 'region' ? -122 : -55,
  }

  const options = {
    series: [
      {
        name: 'Total CO2e',
        data: pageData.data,
      },
    ],
    colors: customBarColors,
    chart: {
      type: 'bar',
      toolbar: {
        offsetX: toolbarOffset.x,
        offsetY: toolbarOffset.y,
        tools: {
          download: `
            <div class="apexcharts-menu-icon" title="Menu">
                <svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true">
                <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"></path>
                </svg>
            </div>   
           `,
        },
      },
    },
    grid: {
      show: false,
      yaxis: {
        lines: {
          show: false,
        },
      },
      xaxis: {
        lines: {
          show: false,
        },
      },
      padding: {
        left: 32,
      },
    },
    plotOptions: {
      bar: {
        horizontal: true,
        barHeight: `${7 * pageData.data.length}%`,
        distributed: true,
      },
    },
    legend: {
      show: false,
    },
    dataLabels: {
      enabled: true,
      textAnchor: 'start',
      formatter: function (_: number, opts: { dataPointIndex: number }) {
        const currentCO2E =
          dataEntries[pageData.page * pageSize + opts.dataPointIndex]?.y

        const formattedPercentage = (currentCO2E / totalCO2EByDataType) * 100
        return formattedPercentage < 0.01
          ? '< 0.01 %'
          : `${formattedPercentage.toFixed(2)} %`
      },
      offsetX: 16,
      background: {
        enabled: true,
        foreColor: mainTheme,
        borderColor: darkTheme,
        padding: 6,
        borderRadius: 1,
        borderWidth: 1,
        opacity: 0.9,
      },
    },
    xaxis: {
      type: 'category',
      labels: {
        show: false,
      },
      axisBorder: {
        show: false,
      },
      max: maxThreshold,
    },
    yaxis: {
      labels: {
        style: {
          fontSize: '13px',
        },
      },
    },
    tooltip: {
      fillSeriesColor: false,
      x: {
        show: false,
      },
      y: {
        formatter: function (value: number, opts: { dataPointIndex: number }) {
          return `${dataEntries[
            pageData.page * pageSize + opts.dataPointIndex
          ].y.toFixed(3)} metric tons`
        },
      },
    },
    height: '500px',
  }

  const handlePage = (page: Page<Entry>) => {
    setPageData(page)
  }

  return (
    <div
      style={{
        minHeight: '500px',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center',
        alignContent: 'center',
      }}
    >
      {emissionsData?.length ? (
        <Fragment>
          {dataType === 'region' && (
            <ChartLegend
              startLabel="Low carbon intensity"
              endLabel="High carbon intensity"
              colorRange={chartBarCustomColors}
            />
          )}
          <Chart
            options={options}
            series={options.series}
            type="bar"
            height={options.height}
          />
        </Fragment>
      ) : (
        <NoDataPage isTop={false} />
      )}
      <div style={{ paddingTop: '10px' }}>
        <Pagination
          data={mappedDataEntries}
          pageSize={pageSize}
          handlePage={handlePage}
        />
      </div>
    </div>
  )
}

export const createCustomBarColors = (
  pageData: Page<Entry>,
  emissionsData: EmissionsRatios[],
  mainTheme: string,
): string[] => {
  const regionColorsMap: string[] = []
  pageData.data.forEach((region) => {
    const currentRegion = region.x
    let color = chartBarCustomColors[0]
    const regionEmissionData = emissionsData.find(
      (item) => item.region === currentRegion,
    ) as EmissionsRatios
    if (!regionEmissionData) {
      regionColorsMap.push(mainTheme)
    } else {
      const { mtPerKwHour } = regionEmissionData
      if (mtPerKwHour >= 0.00064) {
        color = chartBarCustomColors[4]
      } else if (mtPerKwHour >= 0.00048 && mtPerKwHour < 0.00064) {
        color = chartBarCustomColors[3]
      } else if (mtPerKwHour >= 0.00032 && mtPerKwHour < 0.00048) {
        color = chartBarCustomColors[2]
      } else if (mtPerKwHour >= 0.00016 && mtPerKwHour < 0.00032) {
        color = chartBarCustomColors[1]
      } else {
        color = chartBarCustomColors[0]
      }
      regionColorsMap.push(color)
    }
  })
  return regionColorsMap
}
