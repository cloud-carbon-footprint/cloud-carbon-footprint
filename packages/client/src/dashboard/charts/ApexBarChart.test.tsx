/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React from 'react'
import { act, create, ReactTestRenderer } from 'react-test-renderer'
import Chart from 'react-apexcharts'
import moment from 'moment'

import {
  EstimationResult,
  ServiceResult,
  EmissionsRatios,
} from '../../models/types'
import { ApexBarChart, Entry, createCustomBarColors } from './ApexBarChart'
import { Page, Pagination } from './Pagination'
import useRemoteEmissionService from '../client/EmissionFactorServiceHook'
import { fakeEmissionFactors } from '../../data/generateEstimations'

jest.mock('../client/EmissionFactorServiceHook')

const mockedUseEmissionFactorService = useRemoteEmissionService as jest.MockedFunction<
  typeof useRemoteEmissionService
>

describe('ApexBarChart', () => {
  let fixture: ReactTestRenderer
  const data: EstimationResult[] = [
    {
      timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
      serviceEstimates: [
        {
          cloudProvider: 'AWS',
          accountName: 'some account',
          serviceName: 'ebs',
          kilowattHours: 0,
          co2e: 3000.014,
          cost: 0,
          region: 'us-west-1',
        },
        {
          cloudProvider: 'AWS',
          accountName: 'some account',
          serviceName: 's3',
          kilowattHours: 0,
          co2e: 1000.014,
          cost: 0,
          region: 'us-west-2',
        },
        {
          cloudProvider: 'AWS',
          accountName: 'some account',
          serviceName: 'ec2',
          kilowattHours: 0,
          co2e: 2000.014,
          cost: 0,
          region: 'us-west-3',
        },
        {
          cloudProvider: 'AWS',
          accountName: 'some account',
          serviceName: 'eks',
          kilowattHours: 0,
          co2e: 0.000014,
          cost: 0,
          region: 'us-west-4',
        },
      ],
    },
  ]
  beforeEach(() => {
    const mockReturnValue: ServiceResult<EmissionsRatios> = {
      loading: false,
      data: fakeEmissionFactors,
    }
    mockedUseEmissionFactorService.mockReturnValue(mockReturnValue)
    fixture = create(<ApexBarChart data={data} dataType="service" />)
  })
  it('renders with correct configuration', () => {
    expect(fixture.toJSON()).toMatchSnapshot()
  })

  afterEach(() => {
    fixture.unmount()
    mockedUseEmissionFactorService.mockClear()
  })

  it('should format tool tip values with proper data instead of scaled down data', () => {
    const handlePage: (page: Page<Entry>) => void = fixture.root.findByType(
      Pagination,
    ).props?.handlePage
    // make pagination send first page
    act(() => {
      handlePage({
        data: [
          { x: ['ebs', '(AWS)'], y: 100 },
          { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
          { x: ['s3', '(AWS)'], y: 34.00030769056555 },
          { x: ['eks', '(AWS)'], y: 1 },
        ],
        page: 0,
      })
    })

    const yFormatter = fixture.root.findByType(Chart).props?.options?.tooltip?.y
      ?.formatter
    expect(yFormatter).toBeDefined()
    expect(yFormatter(null, { dataPointIndex: 1 })).toEqual(
      '2000.014 metric tons',
    )
  })

  it('should format data label values with proper data instead of scaled down data', () => {
    const handlePage: (page: Page<Entry>) => void = fixture.root.findByType(
      Pagination,
    ).props?.handlePage
    // make pagination send first page
    act(() => {
      handlePage({
        data: [
          { x: ['ebs', '(AWS)'], y: 100 },
          { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
          { x: ['s3', '(AWS)'], y: 34.00030769056555 },
          { x: ['eks', '(AWS)'], y: 1 },
        ],
        page: 0,
      })
    })

    const dataLabelFormatter = fixture.root.findByType(Chart).props?.options
      ?.dataLabels?.formatter
    expect(dataLabelFormatter).toBeDefined()
    expect(dataLabelFormatter(null, { dataPointIndex: 1 })).toEqual('33.33 %')
  })

  it('should format data label values that are less than 0.01', () => {
    const handlePage: (page: Page<Entry>) => void = fixture.root.findByType(
      Pagination,
    ).props?.handlePage
    // make pagination send first page
    act(() => {
      handlePage({
        data: [
          { x: ['ebs', '(AWS)'], y: 100 },
          { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
          { x: ['s3', '(AWS)'], y: 34.00030769056555 },
          { x: ['eks', '(AWS)'], y: 1 },
        ],
        page: 0,
      })
    })

    const dataLabelFormatter = fixture.root.findByType(Chart).props?.options
      ?.dataLabels?.formatter
    expect(dataLabelFormatter).toBeDefined()
    expect(dataLabelFormatter(null, { dataPointIndex: 3 })).toEqual('< 0.01 %')
  })

  it('should filter, sort, order, and scale down data before passing it to Pagination component', function () {
    const paginationComponent = fixture.root.findByType(Pagination)
    const sortedData = [
      { x: ['ebs', '(AWS)'], y: 100 },
      { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
      { x: ['s3', '(AWS)'], y: 34.00030769056555 },
      { x: ['eks', '(AWS)'], y: 1 },
    ]

    expect(paginationComponent.props.data).toEqual(sortedData)
  })

  it('should set colors to each bar based on region emissions', () => {
    fixture = create(<ApexBarChart data={data} dataType="region" />)
    expect(fixture.toJSON()).toMatchSnapshot()
    const paginationComponent = fixture.root.findByType(Pagination)
    const sortedData = [
      { x: ['us-west-1', '(AWS)'], y: 100 },
      { x: ['us-west-3', '(AWS)'], y: 67.00015384528277 },
      { x: ['us-west-2', '(AWS)'], y: 34.00030769056555 },
      { x: ['us-west-4', '(AWS)'], y: 1 },
    ]

    const customColors = ['#790000', '#D99200', '#DF5200', '#00791E']

    expect(paginationComponent.props.data).toEqual(sortedData)

    const optionColors = fixture.root.findByType(Chart).props?.options?.colors
    expect(optionColors).toBeDefined()
    expect(optionColors).toEqual(customColors)
  })

  it('should create custom colors array', () => {
    const firstPagedata = [
      { x: ['us-west-1', '(AWS)'], y: 100 },
      { x: ['us-west-3', '(AWS)'], y: 67.00015384528277 },
      { x: ['us-west-2', '(AWS)'], y: 34.00030769056555 },
      { x: ['us-west-4', '(AWS)'], y: 1 },
      { x: ['us-east-1', '(AWS)'], y: 2 },
    ]
    const pageData: Page<Entry> = { data: firstPagedata, page: 0 }
    const mainTheme = '#2C82BE'

    const emissionsData: EmissionsRatios[] = fakeEmissionFactors
    const colors = ['#790000', '#D99200', '#DF5200', '#00791E', '#73B500']
    const defaultColors = [
      '#2C82BE',
      '#2C82BE',
      '#2C82BE',
      '#2C82BE',
      '#2C82BE',
    ]

    const customColors = createCustomBarColors(
      pageData,
      emissionsData,
      mainTheme,
    )

    expect(customColors).toEqual(colors)

    const mainColors = createCustomBarColors(pageData, [], mainTheme)

    expect(mainColors).toEqual(defaultColors)
  })
})
