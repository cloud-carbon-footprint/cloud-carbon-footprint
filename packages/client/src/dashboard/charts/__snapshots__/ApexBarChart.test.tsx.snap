// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApexBarChart renders with correct configuration 1`] = `
<div
  style={
    Object {
      "alignContent": "center",
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
      "minHeight": "500px",
    }
  }
>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiBox-root MuiBox-root-4"
      data-testid="no-data-page"
    >
      <div
        className="MuiGrid-root MuiGrid-container"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <img
            src="V1Empty-state-generic-icon.svg"
          />
          <div
            className="makeStyles-addSpacing-3"
          >
            There's no data to display!
          </div>
          <div>
            Expand your search parameters to get started.
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-paginationContainer-5"
  >
    <span
      style={
        Object {
          "color": "#ababab",
          "fontWeight": 700,
          "marginRight": "8px",
        }
      }
    >
      1 - 4
       of 
      4
    </span>
    <button
      aria-label="first page"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="chevron-left"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="chevron-right"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="last page"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
